sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala

```diff
@@ -0,0 +1,293 @@
+      for (partitionId <- 0 until numPartitions) {
+        val isLeftSkew = isSkewed(leftStats, partitionId, leftMedSize)
+        val isRightSkew = isSkewed(rightStats, partitionId, rightMedSize)
+        val leftMapIdStartIndices = if (isLeftSkew && supportSplitOnLeftPartition(joinType)) {
+          getMapStartIndices(left, partitionId)
+        } else {
+          Array(0)
+        }
+        val rightMapIdStartIndices = if (isRightSkew && supportSplitOnRightPartition(joinType)) {
+          getMapStartIndices(right, partitionId)
+        } else {
+          Array(0)
+        }
+
+        if (leftMapIdStartIndices.length > 1 || rightMapIdStartIndices.length > 1) {
+          skewedPartitions += partitionId
+          for (i <- 0 until leftMapIdStartIndices.length;
+               j <- 0 until rightMapIdStartIndices.length) {
+            val leftEndMapId = if (i == leftMapIdStartIndices.length - 1) {
+              getNumMappers(left)
+            } else {
+              leftMapIdStartIndices(i + 1)
+            }
+            val rightEndMapId = if (j == rightMapIdStartIndices.length - 1) {
+              getNumMappers(right)
+            } else {
+              rightMapIdStartIndices(j + 1)
+            }
+            // TODO: we may can optimize the sort merge join to broad cast join after
+            //       obtaining the raw data size of per partition,
+            val leftSkewedReader = SkewedPartitionReaderExec(
+              left, partitionId, leftMapIdStartIndices(i), leftEndMapId)
+            val rightSkewedReader = SkewedPartitionReaderExec(right, partitionId,
+              rightMapIdStartIndices(j), rightEndMapId)
+            subJoins += SortMergeJoinExec(leftKeys, rightKeys, joinType, condition,
+              s1.copy(child = leftSkewedReader), s2.copy(child = rightSkewedReader))
+          }
+        }
+      }
+      logDebug(s"number of skewed partitions is ${skewedPartitions.size}")
+      if (skewedPartitions.nonEmpty) {
+        val optimizedSmj = smj.transformDown {
+          case sort @ SortExec(_, _, shuffleStage: ShuffleQueryStageExec, _) =>
+            sort.copy(child = PartialShuffleReaderExec(shuffleStage, skewedPartitions.toSet))
+        }
+        subJoins += optimizedSmj
+        UnionExec(subJoins)
+      } else {
+        smj
+      }
+  }
+
+  override def apply(plan: SparkPlan): SparkPlan = {
+    if (!conf.getConf(SQLConf.ADAPTIVE_EXECUTION_SKEWED_JOIN_ENABLED)) {
+      return plan
+    }
+
+    def collectShuffleStages(plan: SparkPlan): Seq[ShuffleQueryStageExec] = plan match {
+      case _: LocalShuffleReaderExec => Nil
+      case _: CoalescedShuffleReaderExec => Nil
+      case stage: ShuffleQueryStageExec => Seq(stage)
+      case _ => plan.children.flatMap(collectShuffleStages)
+    }
+
+    val shuffleStages = collectShuffleStages(plan)
+
+    if (shuffleStages.length == 2) {
+      // Currently we only support handling skewed join for 2 table join.
```

Comment:
Is this being planned?

Reply:
Is it that we want to avoid things like:
```
SMJ
  SMJ (w/ or w/o Sort)
    Shuffle (w/ or w/o Sort)
    Shuffle (w/ or w/o Sort)
  Shuffle (w/ or w/o Sort)
```
i.e., A SMJ that contains a child SMJ without a Shuffle on top, so we don't wanna optimize the child SMJ coz it changes the outputPartitioning? If so, can we make that clearer in the comment?

Reply:
and since it could affect the outputPartitioning, we need to check if it breaks the operators above.

Reply:
There are scenario's where we can do this for multiple joins right? `INNER` for example.

Reply:
@maryannxue yes, we need to consider the effect of outputPartitioning. And I will updated in the follwing PRs.